examples/contrib/forecast/bart.py

```diff
@@ -0,0 +1,149 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+import argparse
+import logging
+
+import numpy as np
+import torch
+
+import pyro
+import pyro.distributions as dist
+from pyro.contrib.examples.bart import load_bart_od
+from pyro.contrib.forecast import ForecastingModel, backtest
+from pyro.ops.tensor_utils import periodic_cumsum, periodic_repeat
+
+logging.getLogger("pyro").setLevel(logging.DEBUG)
+logging.getLogger("pyro").handlers[0].setLevel(logging.DEBUG)
+
+
+def preprocess(args):
+    """
+    Extract a tensor of (arrivals,departures) to Embarcadero station.
+    """
+    print("Loading data")
+    dataset = load_bart_od()
+    i = dataset["stations"].index("EMBR")
+    arrivals = dataset["counts"][:, :, i].sum(-1)
+    departures = dataset["counts"][:, i, :].sum(-1)
+    data = torch.stack([arrivals, departures], dim=-1)
+    covariates = torch.zeros(len(data), 0)  # Has correct len() but no elements.
+    return data, covariates
+
+
+# We define a model by subclassing the ForecastingModel class and implementing
+# a single .model() method.
+class Model(ForecastingModel):
+    # The .model() method inputs two tensors: a fake tensor zero_data that is
+    # the same size and dtype as the real data (but of course the generative
+    # model shouldn't depend on the value of the data it generates!), and a
+    # tensor of covariates. Our simple model depends on no covariates, so we
+    # simply pass in an empty tensor (see  the preprocess() function above).
+    def model(self, zero_data, covariates):
+        duration = zero_data.size(-2)
+        period = 24 * 7
+
+        # Sample global parameters.
+        noise_scale = pyro.sample("noise_scale",
```

Comment:
this suggests to me that it'd be nice if the default guide could also take a MAP arg that demotes random variables to have dirac delta variational distributions

Reply:
I've intentionally avoided `AutoDelta` and MAP guides because that don't play well with reparameterization. In particular I believe they are incorrect / weird for `Stable` reparameterizers, and I'd like to avoid silent errors in model specification, esp. for a distribution people already distrust.

Reply:
is this even a problem for the globals? shouldn't they be fine?

Reply:
The problem I'm concerned with doesn't depend on local-vs-global; it is due to `AutoDelta(model)` behaving differently from `AutoDelta(poutine.reparam(model, ...))` even for trivial single-site models.

But maybe I misunderstand your concern. Do you simply want to be non-variational about some variables? If so you can use `pyro.param` rather than `pyro.sample` in the model. In case of globals I doubt the prior matters much anyway. I'm only avoiding `pyro.param` here as an academic exercise to demonstrate we can be fully Bayesian.

Reply:
just trying to understand the design constraints---thanks for explaining. 

my question is motivated by the intuition that i'd generally expect being bayesian about noise scale parameters (for example ) using VI could be problematic. so in practice the probably more sensible alternatives are MLE (`param`) and MAP (`AutoDelta`)